### ISSUE TITLE
internal error unexpected BoundT implementing polymorphic interface

### ISSUE THREAD

**Issue by strangemonad** [facebook/flow#3394]
Maybe I'm missing something obvious but I can't seem to make a polymorphic implementation of a polymorphic interface. The simplest case I can reduce it to is the following (es6 syntax):

```
// file foo.js
interface Foo<T> {
  get(): T
}

class FooImpl<T> implements Foo<T> {
  foo: T;

  get(): T {
    return this.foo;
  }
}


// flow error
foo.js:0
Internal error: Failure("Did not expect BoundT")
```

There doesn't seem to be any permutation of the above that allows introducing a type variable and keep the implementation polymorphic in `T` (event if you try and give it an existential `*` type bound).

This of course works fine when you delay unifying the type bound but you loose the early type check ing that `FooImpl` conforms to `Foo` e.g.:
```
interface Foo<T> {
  get(): T
}

class FooImpl<T> {
  foo: T;

  constructor(foo: T) {
    this.foo = foo;
  }

  get(): T {
    return this.foo;
  }
}

const f: Foo<string> = new FooImpl("hello");
```

**Comment by jesseschalken** [comment]
Your sample works in 0.39, but not 0.38 ([link](https://flowtype.org/try/#0JYOwLgpgTgZghgYwgAgGIHt0B4AqA+ZAbwChlkBzCMACgEoAuZHYgX2OIQBs4BnHtTAEkAtgAdOuAsDGcIwiOH4Zs+IqWQxMjHAG52ZSjQZM1ZMlCoBXKCGRgAFsB4A6Tej1k2bIA)). Try upgrading?

**Comment by strangemonad** [comment]
Sweet. I had just upgraded to .38 and missed that .39 was out. I'll try it out and update the issue

**Comment by jesseschalken** [comment]
Duplicate of #3243

**Comment by strangemonad** [comment]
confirmed this is fixed in 0.40
